### PR TITLE
refactor: optimize code with modern Go patterns and pre-allocation

### PR DESCRIPTION
**Pull request by andaaron**
This commit modernizes code across multiple packages by:
- Using Go 1.18+ features (slices.IndexFunc, strings.Cut)
- Pre-allocating slices and maps with known capacity
- Consolidating defensive checks and improving code clarity
- Fixing test data and build tag issues

CLI client improvements:
- Pre-allocate slices in search functions and service methods
- Replace strings.Split with strings. Cut for username:password parsing
- Use range-based iteration instead of manual index loops

Search extension optimizations:
- Cache sort functions in pagination modules
- Pre-allocate page buffers and maps
- Consolidate defensive checks in filterBaseImages/filterDerivedImages
- Fix image base and derived logic allowing out-of-sequence layers for base images
- Fix image pagination reporting images groupped by repos when sorted by update time
- Remove duplicate resolver_test.go file

Monitoring extension:
- Replace manual loops with slices.IndexFunc
- Pre-allocate bucketsFloat2String map

Sync extension:
- Pre-allocate slice in parseRegistryURLs

Test utilities:
- Fix build tags in oci_layout.go

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cli/client/utils.go [hunk 1]
```diff
@@ -93,9 +93,9 @@ func collectResults(config SearchConfig, wg *sync.WaitGroup, imageErr chan strin
 
 func getUsernameAndPassword(user string) (string, string) {
 	if strings.Contains(user, ":") {
-		split := strings.Split(user, ":")
+		username, password, _ := strings.Cut(user, ":")
```
Comment (rchincha) [inline pkg/cli/client/utils.go:96]: what if we have a incorrect entry of "a:b:c:d"?
should we ignore and throw an error, or just accept as before?

File: pkg/extensions/search/resolver_test.go [hunk 2]
```diff
@@ -2396,6 +2395,144 @@ func TestMockedDerivedImageList(t *testing.T) {
 			So(images.Results, ShouldNotBeEmpty)
 			So(len(images.Results), ShouldEqual, 1)
 		})
+
+		Convey("filterDerivedImages edge cases", func() {
```
Comment (rchincha) [inline pkg/extensions/search/resolver_test.go:2399]: are we adding new test cases?

Comment (andaaron) [reply pkg/extensions/search/resolver_test.go:2399]: Coverage for the new error checks.

File: pkg/cli/client/utils.go (hunk 1 above)
Comment (andaaron) [reply pkg/cli/client/utils.go:96]: It's the same to me.

**Review by rchincha — APPROVED** [review]
